test-data/unit/check-optional.test

```diff
@@ -175,3 +175,33 @@ def f(x: None) -> str: pass
 def f(x: int) -> int: pass
 reveal_type(f(None))  # E: Revealed type is 'builtins.str'
 reveal_type(f(0))  # E: Revealed type is 'builtins.int'
+
+[case testOptionalTypeOrTypePlain]
+from typing import Optional
+def f(a: Optional[int]) -> int:
```

Comment:
Done; I've also added special-casing for `None or T` (see tests for what I
mean).